Mentors list on mentors page

Asafailo opened the pull request.

Commit:
Add mentor profile functionalities and avatar support

Introduced mentor-related features, including profile with title and description. Added avatar handling with support for media library integration. Updated database migrations, seeders, and frontend components to enable mentors listing and pagination.

Commit:
Added footer

Commit:
Added footer

Commit:
Refactor codebase for consistency and remove unused dependencies

Updated various files to improve code readability, formatting, and adherence to stricter type declarations. Removed unused imports and cleaned commented-out code for brevity. Added new updates to dependencies in `composer.json` and `yarn.lock`, ensuring consistency across the project.

AratKruglik commented on database/migrations/2025_06_08_152329_add_mentor_description_and_title_to_user_profiles_table.php:
```diff
@@ -0,0 +1,30 @@
+    public function up(): void
+    {
+        Schema::table('user_profiles', function (Blueprint $table): void {
+            $table->string('title', 100)->nullable();
```
А це що за тайтл? ))

Asafailo replied:
Сорі, не писав за це до проходження тестів.  Подумав для карточки ментора треба якийсь  його опис. 
PHP Vue developer/expert, 5 years experience,  
Backend develop
Щось коротке, що можна використати і в seo для пошуку. Якщо то зайве - то я приберу

Commit:
Refactor code and improve test coverage for User resources.

Reordered traits in the User model for consistency and corrected annotation format. Enhanced WelcomePageTest by seeding roles and improving assertion specificity. Minor adjustments to UserResource for better readability.

Commit:
Add comprehensive tests and enhance UserResource handling

Introduces unit tests for `UserResource` to ensure accurate data transformation and profile integration. Updates `WelcomePage` to specify the guard for mentor roles and adjusts related tests. Adds `title` to `UserProfileTest` attribute checks.

Commit:
Add 'meta' key to UserResource response

Updated the UserResource to include an empty 'meta' key in its response. Added a corresponding test in UserResourceTest to verify the presence of the 'meta' key. Removed an unnecessary blank line from the User model's docblock for cleanup.

Commit:
апдейт тестів для mutation coverage

Commit:
Set explicit void return type for user profile tests

Added `: void` return type to test methods in UserProfileTest to ensure explicit type declaration. This change improves code clarity and adheres to stricter type-checking practices.

zloykotuk commented on app/Http/Resources/UserResource.php:
```diff
@@ -0,0 +1,44 @@
+use Illuminate\Http\Resources\Json\ResourceCollection;
+use Override;
+
+class UserResource extends ResourceCollection
```
Якщо ти створюєш resource ти повинен наслідувати клас від JsonResource

Якщо ти передаєш в resource не одне значення а декілька, то потрібно використати метод UserResource::collection(User::all());

Детальніше можеш почитати https://laravel.com/docs/12.x/eloquent-resources

zloykotuk commented on app/Http/Resources/UserResource.php:
```diff
@@ -0,0 +1,44 @@
+                'email'             => $user->email,
+                'created_at'        => $user->created_at,
+                'updated_at'        => $user->updated_at,
+                'profile'           => $user->profile ? [
```
Ти можеш створити для профілю окремий resource щоб потім можна було його перевикористати

zloykotuk commented on app/Actions/Pages/WelcomePage.php:
```diff
@@ -21,6 +23,9 @@ public function handle(): Response
             'phpVersion'     => PHP_VERSION,
+            'mentors'        => new UserResource(User::query()
+                ->role('mentor', 'web')
+                ->with(['profile', 'profile.media'])->paginate(5)),
```
Я трошки не зрозумів чому витягнуємо тільки 5 перших значень, а з решта що ми будемо робити ?

AratKruglik replied:
Ми тут пагінуємо по 5 меторів. Питання чому лише 5 )

AratKruglik replied:
Залишимо ресурс, але там він буде конфліктувати з ресурсом Василя...

Asafailo replied:
додавати фільтр на сторінку чи просто адаптувати по розмірам?

AratKruglik replied:
Фільтри ми додамо потім. Поки всіх скопом, але пагіновано.

maks-oleksyuk commented on composer.json:
```diff
@@ -37,6 +37,7 @@
         "nunomaduro/collision": "^8.8",
         "pestphp/pest": "^3.8",
         "pestphp/pest-plugin-laravel": "^3.1",
+        "rector/rector": "^2.0",
```
це тягнеться з пакета `laravel-reector`, непотрібний рядок

AratKruglik commented on database/factories/UserProfileFactory.php:
```diff
@@ -23,14 +22,14 @@ class UserProfileFactory extends Factory
     public function definition(): array
     {
         return [
-            'user_id'     => User::factory(),
```
Не потрібно прибирати це поле. Фабрика буде використовуватись для тестів. Це поле обовʼязкове.

Asafailo replied:
була проблема що генерувалось два профіля. 
Один сідером, інший запускав observer, який створював додатковий профіль до поточного, подумав що варто розділити окремо сідер під юзера, окремо під профіль. поправлю

AratKruglik replied:
Ну так то вже треба в сідері регулювати що створювати, а що - ні. Фабрика тут ні до чого ж наче.

AratKruglik commented on database/seeders/UserSeeder.php:
```diff
@@ -14,9 +17,17 @@ class UserSeeder extends Seeder
+            ->count(10)
+            ->create()
+            ->each(function ($user): void {
+                $user->assignRole(RoleEnum::MENTOR->value, RoleGuardEnum::MENTOR->value);
```
Сорян за питання, звісно, а чому ми саме в цьому сідері генеруємо менторів? :)

Asafailo replied:
погоджуюсь, нелогічно.  Як правильніше - зробити окрему генерацію по кожній ролі в одному сідері? 10 менторів, 10 звичайних юзерів, 10 менті, 10 коучів..?

AratKruglik replied:
Краще зробити генерацію по ролях в коремих сідерах, щоб мати можливість точково застосовувати їх саме тоді, коли це потрібно.

AratKruglik commented on resources/js/Components/Navigation/Footer.vue:
```diff
@@ -1,54 +1,62 @@
-              </ul>
+    <footer class="relative mx-auto mt-32 max-w-7xl px-6 lg:px-8">
+        <div class="justify-center mt-4 items-center">
+            <MentorsList/>
```
Що у футері робить список менторів? О_о

Asafailo replied:
неправильно зрозумів задачу, думав це як на сайтах в нижній секції працівників відображають іноді. 
поправлю

AratKruglik commented on tests/Feature/Pages/WelcomePageTest.php:
```diff
@@ -3,12 +3,27 @@
+    ]);
+    $user->syncRoles($role);
 
+    $response = $this->get('/');
```
Давай використовувати `route({routeName})` для тестів замість хардкоду URI

AratKruglik commented on tests/Unit/Actions/Pages/WelcomePageTest.php:
```diff
@@ -32,11 +40,71 @@
+            $user->syncRoles($role);
+        });
+
+        $response = $this->get('/');
```
не хардкодимо URI

AratKruglik commented on tests/Unit/Http/Resources/UserResourceTest.php:
```diff
@@ -0,0 +1,153 @@
+            ->toBeArray()
+            ->toHaveKey('data')
+            ->and($result['data'])
+            ->toBeInstanceOf(Illuminate\Support\Collection::class)
```
винеси в use цей неймспейс, будь ласка

AratKruglik reviewed (commented):
З глобального - прибираємо UserResource

Commit:
Remove unused tests and refactor mentor resource handling

Deleted obsolete `UserResourceTest` while improving mentor-related resource handling across the application. Key updates include refactoring avatar defaults, enhancing `UserSeeder` functionality, and consolidating mentor data processing in the `WelcomePage` flow.

Commit:
Remove 'avatar' from appended attributes

The 'avatar' attribute has been removed from the `$appends` property to streamline the UserProfile model. This ensures unnecessary data transformations are avoided and improves model performance.

Commit:
Refactor avatar handling in UserProfile model

Updated TEST_AVATAR_URL to accept dynamic names and added the `avatar` property to the `$appends` array for easier access. Also included the `avatar` URL property in the PHPDoc for clarification.

Commit:
Add missing annotations for IDE helper mixins

Updated the User and UserProfile models to include `@mixin` annotations for IDE helper functionality. This ensures better autocomplete support and improves development efficiency.

AratKruglik commented on database/seeders/UserSeeder.php:
```diff
@@ -5,18 +5,46 @@
+
+    public function __construct()
+    {
+        $this->roles = Role::query()->get();
```
У тебе є Enum з ролями. не бачу сенсу лізти в базу з ролями. 
Якщо потрібно взяти рандомну роль, то `fake()->randomElement()` в поміч.

Asafailo replied:
в поточному сідері по ролях - в нас немає записів з гардом Web для всіх ролей.
 По замовчуванню при асайні ролі підставляє web. 

>   Spatie\Permission\Exceptions\RoleDoesNotExist 
  There is no role named `mentor` for guard `web`.

як правильно зробити
1) додати в сідер для кожної ролі(mentor, menti, user...) окремий запис з гардом веб? 
2) Чи все ж таки діставати роль з бази

AratKruglik replied:
Ця помилка виникає в тестах? Якщо так, то в тестах перед виконанням тесту варто викликати наступний метод `$this->seed(RoleSeeder::class);`

Asafailo replied:
ні, я запускаю сідер перед тестами. і окремо в контейнері теж запускав. 
Питання в тому що сідер створює записи в таблиці roles. 

name    ||  guard
____________ 
user      ||   user
mentor ||  mentor 
menti    || menti
....
при цьому немає

немає 
user   web
mentor web 

коли я намагаюсь заасайнити на юзера роль наприклад mentor - з'являється помилка що не існує ролі mentor для guard web. 

Я не дуже розумію загальну логіку, яка буде в плані авторизації.  можливо треба конфіги auth розширити. 

Але вирішення проблеми було - або відразу асайнити на роль умовно роль Mentor, guard Mentor. для цього звертався до колекції.  Або ж постворювати ролі mentor => web, і тут мені не дуже зрозуміло чи це ок.

AratKruglik replied:
тоді варто розібратись чому ролі чіпають гарду `web` і як то зарезолвити окрім додавання гардів для ролей в базу. Думаю так буде правильніше.

Asafailo replied:
добре, повернусь

Asafailo replied:
подивився - нажаль не знайшов інших варіантів. 
Spatie permission вимагає створення пари role - guard для кожного використовуємого гарда в ролі. 

А по дефолту - якщо завантажити тільки назву ролі в assignRole - то підставляється перший дефолтний гард з auth. в даному випадку web.  тому або використовувати модель Role - з вже створених сідером.  mentor\mentor наприклад. 
Або створити копії ролей для гарду web, але це може створити плутанину потенційно. 

зміна першого дефолтного гарда в auth - нічого не змінює - тому що просто буде братись перший. 

І я розділив сідери по кожній ролі, але поки з однаковою логікою. 

Ще момент - кожен юзер крім ролі ментор. менті... зараз має і роль юзера як загальну. через UserObserver. якщо це зайве - дайте знати 

https://spatie.be/docs/laravel-permission/v6/basic-usage/multiple-guards

AratKruglik commented on resources/js/Layouts/GuestLayout.vue:
```diff
@@ -9,14 +10,17 @@ import { Link } from '@inertiajs/vue3';
         <div>
             <Link href="/">
-                <ApplicationLogo class="h-20 w-20 fill-current text-gray-500" />
+                <ApplicationLogo class="h-18 w-18 fill-current text-gray-500"/>
```
Чому змінюємо ширину блоку? Чи не відобразиться це потім на інших сторінках?

AratKruglik commented on resources/js/Layouts/GuestLayout.vue:
```diff
@@ -9,14 +10,17 @@ import { Link } from '@inertiajs/vue3';
 
         <div
-            class="mt-6 w-full overflow-hidden bg-white px-6 py-4 shadow-md sm:max-w-md sm:rounded-lg"
+            class="mt-4 w-full overflow-hidden bg-white px-6 py-4 shadow-md sm:max-w-md sm:rounded-lg"
```
Те саме питання, що вище

Commit:
Add new seeders and roles for user types and UI adjustments

This commit introduces seeders for Admin, Menti, Coach, and SuperAdmin along with their associated roles and guards in the auth config. UI adjustments were made to the GuestLayout component, including logo size and padding updates. Middleware aliases were also added to support role and permission checks.

Commit:
Remove obsolete UserResource and middleware aliases.

The UserResource and related middleware aliases were no longer in use and have been removed to clean up the codebase. This simplifies the application and eliminates unused code.

Commit:
Add MentorProgramSeeder and CurrencySeeder to seeding process

Introduce `MentorProgramSeeder` to populate mentor-related data and `CurrencySeeder` to seed currency information. Updated `DatabaseSeeder` to include the new seeders, ensuring comprehensive database initialization for development and testing.

Commit:
Refactor models, tests and introduce new mentorship features

Removed redundant coding standards file and refactored models for cleaner structure and enhanced relationships. Added new unit tests for mentorship programs, implemented policies for authorization, and introduced mentorship page and review functionalities. Updated testing suite to ensure compliance with new changes.

Commit:
Simplify auth configuration and fix avatar handling.

Removed unused guards in the auth configuration file to streamline setup. Updated tests to properly handle default avatar URLs by using the constant `UserProfile::DEFAULT_AVATAR_URL`. This ensures consistency and reduces reliance on hardcoded placeholders.

maks-oleksyuk commented on database/migrations/2025_06_08_152329_add_mentor_description_and_title_to_user_profiles_table.php:
```diff
@@ -0,0 +1,30 @@
+    public function down(): void
+    {
+        Schema::table('user_profiles', function (Blueprint $table): void {
+            $table->string('title', 100)->nullable();
```
неправильна міграція

AratKruglik replied:
В чому?

maks-oleksyuk replied:
тому що це ролбек міграція

AratKruglik replied:
Упс, провтикав, що це ролбек. Дякую.

AratKruglik replied:
@Asafailo перевір будь ласка )

AratKruglik committed:
Merge develop branch resolving conflicts

Commit:
Add cost_per_hour and currency_id to UserProfile

Extended the UserProfileFactory to include cost_per_hour and currency_id attributes for better data generation in tests. Updated the UserProfile model to append these attributes to its JSON representation.

Commit:
Merge remote-tracking branch 'origin/feature/33-add-list-of-mentors-page' into feature/33-add-list-of-mentors-page

Commit:
Remove 'title' column rollback logic in migration

The `down` method now correctly drops the 'title' column instead of re-adding it. This ensures the migration rollback behaves as expected and maintains database integrity.

Asafailo committed:
Merge branch 'develop' into feature/33-add-list-of-mentors-page

Commit:
Add cost_per_hour and currency_id fields to UserProfile

Extended the UserProfile model and factory to include cost_per_hour and currency_id attributes. Updated mass assignable, visible, and appended properties accordingly to ensure proper handling of these new fields.

Commit:
Merge remote-tracking branch 'origin/feature/33-add-list-of-mentors-page' into feature/33-add-list-of-mentors-page

Commit:
Update dependencies in composer.lock file

Upgraded several packages to newer versions, including `brick/math`, `filament`, `laravel/framework`, and others. This ensures compatibility, feature improvements, and the latest bug fixes for the project.
